### ISSUE TITLE
kv: provide mechanism to ensure currently blocked transactions progress

### ISSUE THREAD

**Issue by ajwerner** [cockroachdb/cockroach#55981]
**Is your feature request related to a problem? Please describe.**

This issue arises in the context of #55047. That proposal has a problem related to deadlocks. The gist of that proposal is that transactional schema changes will require coordinating with concurrent transactions by propagating new descriptor versions and waiting for them to be adopted. An issue that can arise is that an open transaction using an older descriptor version might be blocked on an intent due to the user transaction (or transitively due to some other transaction which is blocked on the user transaction). The polling is effectively the user transaction blocking on concurrent transactions using the current version of the
table descriptor. In order to avoid deadlock, we'll need to augment our deadlock detection protocol with an affordance for this case. 

There are two cases of transactions involved in such a deadlock: reading transactions and writing transaction. Ideally all reading transactions would be allowed to proceed as the hope is that the schema change will push its timestamp above them. Writing transactions, however, must be aborted.

**Describe the solution you'd like**

It is unfortunate to think about involving client logic in deadlock detection as it has thus far remained rather self-contained. The hope in this issue is that we can find a clean interface to represent the logic needed to resolve this complication.

In order to achieve the goals, we'll need some new functionality in controlling the transaction. We'll discuss how we might encapsulate these raw concepts with better APIs afterward.

* Ability to push the transaction to an arbitrary timestamp in the future
    * This is a general requirement for #55047 and likely deserves its own issue.
    * Optionally refresh deferred.
* Ability to observe the type and timestamp of push requests in QueryTxn.

From a high level, the waiting protocol will look like:

* Commit new descriptor version using child transaction
* Drop leases for table modified by child transaction in user transaction
* While true:
    * Scan for outstanding leases with new child txn, if empty, break, recording timestamp at which the read occurred
    * Issues QueryTxn against user transaction
        * If waiters with PushTimestamp, push parent transaction's timestamp above highest (refresh deferred), force a heartbeat
        * If waiters with PushAbort, push abort all of them with high (numeric) priority.
* Push user transaction to the timestamp at which all of the leases were observed to be dropped and refresh
* Re-acquire relevant leases

The above protocol exposes some serious guts of the transactional model. It'd be best to hide most of these details underneath the KV API. How best to do that is not clear. Perhaps we want something like:

```
package kv

// ObserveCondition will call the function with a child transaction of txn. If no error and true is returned,
// the parent transaction will be pushed above the timestamp at which the child transactions commits.
// If false is returned, the call will attempt to eliminate deadlocks and then will back off according to the
// retrier's implied backoff.
func ObserveCondition(ctx context.Context, txn *Txn, retryOpts retry.Options, cond func(ctx context.Context, childTxn *Txn) (done bool, _ error)) error  
```

**Describe alternatives you've considered**

None. 

Jira issue: CRDB-3604

**Comment by ajwerner** [comment]
@nvanbenschoten can you give this a review when you find some time. No rush as this or anything like it will need to be sequenced after we have child transactions.

**Comment by nvanbenschoten** [comment]
> Ability to push the transaction to an arbitrary timestamp in the future

An arbitrary timestamp in the future, or an arbitrary timestamp <= present time?

> Issues QueryTxn against user transaction
> If waiters with PushTimestamp, push parent transaction's timestamp above highest (refresh deferred), force a heartbeat
> If waiters with PushAbort, push abort all of them with high (numeric) priority.

This seems to be the heart of the new protocol. The idea is that we allow all read-write conflicts to succeed by pushing ourselves above their push timestamp and we fail all write-write conflicts by aborting the conflicting transaction. That all seems reasonable, though somewhat complex.

I wonder whether we even strictly need the extra complexity for read-write conflicts. Does aborting any conflict seem too disruptive? Is the concern that it could cause read-only txns that could previously never retry to begin hitting txn retries? If there's read-write contention, then I'm not sure that will ever be the case, because these read-write txns need to be ready to handle ReadWithinUncertaintyInterval errors. I ask because an `AbortConflicts` txn API would be a lot simpler to think about than what's being proposed here.
